consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java

```diff
@@ -12,35 +12,217 @@
+    } else {
+      LOG.info(
+          "Block timer expired for a round ({}) other than current ({})",
+          roundIdentifier,
+          currentRound.getRoundIdentifier());
+    }
+  }
+
+  public void roundExpired(final RoundExpiry expire) {
+    if (!expire.getView().equals(currentRound.getRoundIdentifier())) {
+      LOG.debug("Ignoring Round timer expired which does not match current round.");
+      return;
+    }
+
+    LOG.debug("Round has expired, creating PreparedCertificate and notifying peers.");
+    final Optional<PreparedCertificate> preparedCertificate =
+        currentRound.createPrepareCertificate();
 
-  public void handlePrepareMessage(final SignedData<PreparePayload> prepareMsg) {}
+    if (preparedCertificate.isPresent()) {
+      latestPreparedCertificate = preparedCertificate;
+    }
 
-  public void handleCommitMessage(final SignedData<CommitPayload> commitMsg) {}
+    startNewRound(currentRound.getRoundIdentifier().getRoundNumber() + 1);
 
-  public void handleBlockTimerExpiry(final ConsensusRoundIdentifier roundIndentifier) {}
+    final SignedData<RoundChangePayload> localRoundChange =
+        messageFactory.createSignedRoundChangePayload(
+            currentRound.getRoundIdentifier(), latestPreparedCertificate);
+    transmitter.multicastRoundChange(currentRound.getRoundIdentifier(), latestPreparedCertificate);
 
-  public void handleRoundChangeMessage(final SignedData<RoundChangePayload> roundChangeMsg) {}
+    // Its possible the locally created RoundChange triggers the transmission of a NewRound
+    // message - so it must be handled accordingly.
+    handleRoundChangeMessage(localRoundChange);
+  }
+
+  public void handleProposalMessage(final SignedData<ProposalPayload> msg) {
+    LOG.info("Received a Proposal message.");
+    actionOrBufferMessage(
+        msg,
+        () -> currentRound.handleProposalMessage(msg),
+        (roundstate) -> roundstate.setProposedBlock(msg));
+  }
+
+  public void handlePrepareMessage(final SignedData<PreparePayload> msg) {
+    LOG.info("Received a prepare message.");
+    actionOrBufferMessage(
+        msg,
+        () -> currentRound.handlePrepareMessage(msg),
+        (roundstate) -> roundstate.addPrepareMessage(msg));
+  }
+
+  public void handleCommitMessage(final SignedData<CommitPayload> msg) {
+    LOG.info("Received a commit message.");
+    actionOrBufferMessage(
+        msg,
+        () -> currentRound.handleCommitMessage(msg),
+        (roundstate) -> roundstate.addCommitMessage(msg));
+  }
+
+  private void actionOrBufferMessage(
+      final SignedData<? extends Payload> msg,
+      final Action inRoundHandler,
+      final Consumer<RoundState> buffer) {
+    final ConsensusRoundIdentifier msgRoundId = msg.getPayload().getRoundIdentifier();
+    if (msgRoundId.equals(currentRound.getRoundIdentifier())) {
+      inRoundHandler.apply();
+    } else if (msgRoundId.getRoundNumber() > currentRound.getRoundIdentifier().getRoundNumber()) {
```

Comment:
perhaps extract the msgRound.getRoundNumber() > currentRound.getRoundIdentifier().getRoundNumber as this a common check that being used in many places e.g. in handleRoundChangeMessage and handleNewRoundMessage

Reply:
Done - sort of, introduced the concept of "MessageAge".